test/suite/test_compact05.py

```diff
@@ -0,0 +1,106 @@
+# of the public at large and to the detriment of our heirs and
+# successors. We intend this dedication to be an overt act of
+# relinquishment in perpetuity of all present and future rights to this
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# test_compact05.py
+#   Test compaction through the background compaction server.
+#
+
+import wttest
+from wiredtiger import stat
+from wtscenario import make_scenarios
+
+# Test if compaction proceeds depending on the free_space_target field used by the background
+# compaction server.
+class test_compact05(wttest.WiredTigerTestCase):
+
+    free_space_target = [
+        # The threshold is smaller than the number of available bytes, compaction should proceed.
+        ('1MB', dict(compact_config='1MB', expected_stdout=None)),
+        # The threshold is greater than the number of available bytes, compaction should not proceed.
+        ('45MB', dict(compact_config='45MB', expected_stdout='number of available bytes.*is less than the configured threshold')),
+    ]
+    scenarios = make_scenarios(free_space_target)
+
+    conn_config = 'statistics=(all),verbose=(compact_progress,compact)'
+    create_params = 'key_format=i,value_format=S,allocation_size=4KB,leaf_page_max=32KB,'
+    table_numkv = 100 * 1000
+    table_uri='table:test_compact05'
+
+    delete_range_len = 10 * 1000
+    delete_ranges_count = 4
+    value_size = 1024 # The value should be small enough so that we don't create overflow pages.
+
+    # Create the table in a way that it creates a mostly empty file.
+    def test_compact05(self):
+
+        # Create the table and populate it with a lot of data.
+        self.session.create(self.table_uri, self.create_params)
+        c = self.session.open_cursor(self.table_uri, None)
+        for k in range(self.table_numkv):
+            c[k] = ('%07d' % k) + '_' + 'abcd' * ((self.value_size // 4) - 2)
+        c.close()
+        self.session.checkpoint()
+
+        # Now let's delete a lot of data ranges. Create enough space so that compact runs in more
+        # than one iteration.
+        c = self.session.open_cursor(self.table_uri, None)
+        for r in range(self.delete_ranges_count):
+            start = r * self.table_numkv // self.delete_ranges_count
+            for i in range(self.delete_range_len):
+                c.set_key(start + i)
+                c.remove()
+        c.close()
+
+        # Compact!
+        compact_config = f"free_space_target={self.compact_config}"
+        if self.compact_config == '1MB':
```

Comment:
Minor suggestion: You could have a variable in the dict in the free_space_target array that says whether compaction should happen, so that the test definition logic is all in one place. But either approach is fine.

Reply:
Good idea, I improved a few things along the way in 4b3be6ac3247f9ea16ce52ae5d23216a23e5cabd.